### DISCUSSION TITLE
Generating model types from DB schema (could improve DX)

### DISCUSSION THREAD

**Discussion by harrywebdev** [evoluhq/evolu#58] [Ideas]
Take README example:

```
const { useQuery, useMutation } = createHooks({
  todo: {
    id: model.id<"todo">(),
    title: model.NonEmptyString1000,
    isCompleted: model.SqliteBoolean,
  },
});
```

now when I'd like to have a component, that takes a `todo` record as a prop, I don't have a simple way, how to type it. When querying the DB:

```
const { rows } = useQuery(function selectTodos(db) {
  return db
    .selectFrom("todo")
    .select(["id", "title", "isCompleted"])
    .where("isDeleted", "is not", model.cast(true))
})
```

the type of a single row is (at least according to my IDE):

```
const record: Selection<From<DbSchemaToType<{todo: {id: ZodBranded<ZodEffects<ZodString, string, string>, "todoId">, title: ZodBranded<ZodString, "NonEmptyString1000">, isCompleted: ZodBranded<...>}}, {}>, "todo">, "todo", "title" | ... 1 more ... | "id"> | undefined
```

Could there be a way, how I could get a nice model type from the DB schema? Or maybe a different way, how to specify the model first?

**Comment by steida** [comment]
I was thinking about it and decided not to do that. The reason is schema evolving. Sooner or later, you will add some props and deprecate the other. But after the release, you can't remove deprecated props because there could be data with such deprecated props waiting to be synced. I don't believe some generic model would survive for a longer time. There would be deprecated props.

The right approach is to redeclare the props you are going to use. For example, you rarely will need a full "Todo model" in a listing. Only a few props like ID and title. You probably will not need the full "Todo model" in detail either because you will use joins. 

"duplication is cheaper than the wrong abstraction”, so be explicit and type the props you will use. Feel free to disagree.

Also, check https://github.com/evoluhq/evolu/wiki/Nullability-and-version-less-schema

**Reply by harrywebdev** [reply]
Makes sense actually - it requires quite a different mindset when working with CRDTs.

**Reply by steida** [reply]
@harrywebdev But you were also correct. Evolu 2.0 will use Schema instead of Zod, and types will be nicely reusable. And faster and easier to read too.

**Comment by steida** [comment]
Evolu already has that
